### DISCUSSION TITLE
CopyQ paste doesn't work in terminal (zsh)

### DISCUSSION THREAD

**Discussion by ittaysw** [wezterm/wezterm#2157] [Q&A]
Shift-Insert works, but when I open the copyq menu (which is bound to ctrl-shift-v via a custom gnome shortcut to `copyq menu`), nothing is pasted. This works in Terminal. Using X11.

**Comment by wez** [comment]
There's not enough information here to help you. Please consider spelling out in more detail the steps necessary to recreate the issue.

**Reply by ittayd** [reply]
Since I don't know exactly what the root cause here, I'm trying to detail everything I know of the setup. Repeating here:
* Ubuntu 20.04, regular (not wayland) gnome.
* Using zsh 
* CopyQ installed
* A custom shortcut in Gnome settings maps ctrl-shift-v to the command `copyq menu`
* In wezterm, shift-insert is mapped to PasteFrom="Clipboard"
* Copy some text
* In wezterm, shift-insert will paste that text
* In wezterm, ctrl-shift-v. A copyq menu appears. Select the first entry (the text that was just copied) --> will not paste anything
* In gnome terminal, ctrl-shift-v and selecting the first entry will paste as expected

**Reply by MuhammedZakir** [reply]
Note: I don't use CopyQ.

Reading CopyQ's doc, it uses `shift+insert` for pasting, which by default pastes from `primary` selection in Linux. Try reverting `shift+insert` binding to paste from `primary` selection in WezTerm. That maybe the culprit.

**Reply by ittayd** [reply]
When it was, pasting from the clipboard didn't work in wezterm. And it works in gnome terminal

**Reply by MuhammedZakir** [reply]
It worked in Gnome Terminal because, unlike in WezTerm, you didn't remap `shift+insert` to paste from `clipboard` selection?

Edit: Anyway, we will know whether this is the case if you revert `shift+insert` to default mapping in WezTerm.

**Reply by ittayd** [reply]
Shift-insert in gnome terminal pastes from the clipboard.

**Reply by MuhammedZakir** [reply]
It's working for me. What I did:

1. In WezTerm, bound `shift+insert` to `PasteFrom="Clipboard"`.
2. In CopyQ > Commands/Global Shortcuts: set `ctrl+shift+v` to show tray menu.
3. Pressed `ctrl+shift+v` in WezTerm -> CopyQ tray menu showed up -> activated an entry by pressing enter/return.
4. The activated entry was pasted in WezTerm.

OS: Manjaro XFCE (X11)
WezTerm version: 20220619-214005-be8a6e18

**Reply by ittayd** [reply]
Ok, so how do I go about debugging this in my environment?

**Reply by MuhammedZakir** [reply]
I am not sure.

1. After selecting an entry in CopyQ, check if the content is copied correctly. Execute this in WezTerm:
```zsh
for s in clipboard primary; echo "$s: " && xclip -out -selection $s && echo;
```
2. Are you able to paste using CopyQ with default config?
    - Simple way in ZSH: `wezterm --config-file =(echo return {})`.
3. If yes to #_1_, then reduce the WezTerm config to just remapping `shift+insert`, and try again.
4. Try latest WezTerm nightly version.

**Reply by wez** [reply]
(Note that `wezterm -n` is the most convenient way to run wezterm with the default configuration)

**Reply by ittayd** [reply]
OK, so with `wezterm -n` everything works as expected (using nightly 20220618-195835-2f073a7d). 

Tried `wezterm` with no .wezterm.lue configuration file and copyq doesn't work (shift-insert works, ctrl-insert doesn't... ). 

I noticed that the wezterm icon actually runs wezterm-gui so I tried to run wezterm from gnome terminal and copyq didn't work, but, if i run `wezterm` again from the wezterm instance then copyq works. Super weird. 

The xclip one liner works as expected

**Reply by MuhammedZakir** [reply]
That's wierd. Does WezTerm detect `shift+insert` send by CopyQ? Try:
1. Add `debug_key_events = true` in WezTerm config.
2. Run WezTerm from another terminal (to see the log).
    - Running instances' logs can be seen in `dirname $WEZTERM_UNIX_SOCKET`.
3. Use CopyQ.
4. Check whether WezTerm logs key events for `shift+insert`.

**Reply by ittayd** [reply]
This is what wezterm saw:
```
08:11:52.967  INFO   wezterm_gui::termwindow::keyevent     > key_event RawKeyEvent { key: Physical(LeftShift), modifiers: NONE, phys_code: Some(LeftShift), raw_code: 50, repeat_count: 1, key_is_down: true, handled: Handled(false) } 
08:11:52.967  INFO   wezterm_gui::termwindow::keyevent     > key_event KeyEvent { key: Shift, modifiers: NONE, repeat_count: 1, key_is_down: true, raw: Some(RawKeyEvent { key: Physical(LeftShift), modifiers: NONE, phys_code: Some(LeftShift), raw_code: 50, repeat_count: 1, key_is_down: true, handled: Handled(false) }) } 
08:11:52.967  INFO   wezterm_gui::termwindow::keyevent     > send to pane DOWN key=Shift mods=NONE
08:11:52.968  INFO   wezterm_gui::termwindow::keyevent     > key_event RawKeyEvent { key: Physical(Insert), modifiers: NONE, phys_code: Some(Insert), raw_code: 118, repeat_count: 1, key_is_down: true, handled: Handled(false) } 
08:11:52.968  INFO   wezterm_gui::termwindow::keyevent     > key_event KeyEvent { key: Insert, modifiers: NONE, repeat_count: 1, key_is_down: true, raw: Some(RawKeyEvent { key: Physical(Insert), modifiers: NONE, phys_code: Some(Insert), raw_code: 118, repeat_count: 1, key_is_down: true, handled: Handled(false) }) } 
08:11:52.968  INFO   wezterm_gui::termwindow::keyevent     > send to pane DOWN key=Insert mods=NONE
08:11:52.968  INFO   wezterm_term::terminalstate::keyboard > key_down: sending "\u{1b}[2~", Insert NONE
08:11:52.970  INFO   wezterm_gui::termwindow::keyevent     > key_event KeyEvent { key: Char('\r'), modifiers: NONE, repeat_count: 1, key_is_down: false, raw: Some(RawKeyEvent { key: Physical(Return), modifiers: NONE, phys_code: Some(Return), raw_code: 36, repeat_count: 1, key_is_down: false, handled: Handled(false) }) } 
08:11:52.970  INFO   wezterm_gui::termwindow::keyevent     > send to pane UP key=Enter mods=NONE
08:11:52.970  INFO   wezterm_gui::termwindow::keyevent     > key_event KeyEvent { key: Insert, modifiers: NONE, repeat_count: 1, key_is_down: false, raw: Some(RawKeyEvent { key: Physical(Insert), modifiers: NONE, phys_code: Some(Insert), raw_code: 118, repeat_count: 1, key_is_down: false, handled: Handled(false) }) } 
08:11:52.970  INFO   wezterm_gui::termwindow::keyevent     > send to pane UP key=Insert mods=NONE
08:11:52.971  INFO   wezterm_gui::termwindow::keyevent     > key_event KeyEvent { key: Shift, modifiers: NONE, repeat_count: 1, key_is_down: false, raw: Some(RawKeyEvent { key: Physical(LeftShift), modifiers: NONE, phys_code: Some(LeftShift), raw_code: 50, repeat_count: 1, key_is_down: false, handled: Handled(false) }) } 
08:11:52.971  INFO   wezterm_gui::termwindow::keyevent     > send to pane UP key=Shift mods=NONE

```

**Reply by wez** [reply]
SHIFT doesn't appear to be set in modifiers when Insert is generated, even though we can see that the shift key was pressed immediately prior to that.
Try setting `use_ime = false`.  Over in https://github.com/wez/wezterm/issues/2159 it looks like the IME can munge unusual shift key combinations.

**Reply by MuhammedZakir** [reply]
FYI: in latest nightly, WezTerm will also log triggered actions for key events which will make debugging easier. E.g.
```
10:09:26.592  INFO   wezterm_gui::termwindow::keyevent     > Insert SHIFT -> perform PasteFrom(PrimarySelection)
```

**Reply by ittayd** [reply]
@wez , that made copyq work. but still ctrl-insert doesn't. That is, if I select text in the terminal and copy with ctrl-insert then try to paste here, I get the previously copied text. It works as expected in gnome-terminal

**Reply by MuhammedZakir** [reply]
Presuming you are using `shift+insert` for pasting and it's remapped to paste from `clipboard`, then you will also have to remap `ctrl+insert` to copy to `clipboard` selection. By default, it copies to `primary` selection.

**Reply by ittayd** [reply]
I removed that mapping. No keys are mapped. The question is, should I add a ctrl-insert to copy to the clipboard? Judging from gnome-terminal, it looks like this what it does.

**Reply by MuhammedZakir** [reply]
> The question is, should I add a ctrl-insert to copy to the clipboard? 

If that's what you want, go ahead. Personally, I dislike it because the whole point of having different key combos for copy and paste is to have multiple clipboards. But, I can understand why some would want to use only 1 clipboard; I used to do it too.

> Judging from gnome-terminal, it looks like this what it does.

XFCE Terminal does the same - copies to `clipboard` or maybe both `primary` and `clipboard`.

Edit: If you want to copy to multiple clipboards using `ctrl+insert`, see https://wezfurlong.org/wezterm/config/lua/keyassignment/Multiple.html.
